scene/3d/label_3d.cpp

```diff
@@ -0,0 +1,1049 @@
+				WordCache *wc = memnew(WordCache);
+				if (word_cache) {
+					last->next = wc;
+				} else {
+					word_cache = wc;
+				}
+				last = wc;
+
+				wc->pixel_width = current_word_size;
+				wc->char_pos = word_pos;
+				wc->word_len = i - word_pos;
+				wc->space_count = space_count;
+				current_word_size = 0;
+				space_count = 0;
+			}
+			if (current_word_size == 0) {
+				word_pos = i;
+			}
+			char_width = font->get_char_size(current, xl_text[i + 1]).width;
+			current_word_size += char_width;
+			line_width += char_width;
+
+			// allow autowrap to cut words when they exceed line width
+			if (autowrap && (current_word_size > max_line_width)) {
+				separatable = true;
+			}
+		}
+
+		if ((autowrap && (line_width >= max_line_width) && ((last && last->char_pos >= 0) || separatable)) || insert_newline) {
+			if (separatable) {
+				if (current_word_size > 0) {
+					WordCache *wc = memnew(WordCache);
+					if (word_cache) {
+						last->next = wc;
+					} else {
+						word_cache = wc;
+					}
+					last = wc;
+
+					wc->pixel_width = current_word_size - char_width;
+					wc->char_pos = word_pos;
+					wc->word_len = i - word_pos;
+					wc->space_count = space_count;
+					current_word_size = char_width;
+					word_pos = i;
+				}
+			}
+
+			WordCache *wc = memnew(WordCache);
+			if (word_cache) {
+				last->next = wc;
+			} else {
+				word_cache = wc;
+			}
+			last = wc;
+
+			wc->pixel_width = 0;
+			wc->char_pos = insert_newline ? WordCache::CHAR_NEWLINE : WordCache::CHAR_WRAPLINE;
+
+			line_width = current_word_size;
+			line_count++;
+			space_count = 0;
+		}
+	}
+
+	word_cache_dirty = false;
+}
+
+void Label3D::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			if (!pending_update) {
+				_im_update();
+			}
+		} break;
+		case NOTIFICATION_TRANSLATION_CHANGED: {
+			String new_text = tr(text);
+			if (new_text == xl_text) {
+				return; // Nothing new.
+			}
+			xl_text = new_text;
+
+			regenerate_word_cache();
+			_queue_update();
+		} break;
+	}
+}
+
+void Label3D::_im_update() {
+	_shape();
+
+	triangle_mesh.unref();
+	update_gizmo();
+
+	pending_update = false;
+}
+
+void Label3D::_queue_update() {
+	if (pending_update) {
+		return;
+	}
+
+	pending_update = true;
+	call_deferred(SceneStringNames::get_singleton()->_im_update);
+}
+
+AABB Label3D::get_aabb() const {
+	return aabb;
+}
+
+Ref<TriangleMesh> Label3D::generate_triangle_mesh() const {
+	if (triangle_mesh.is_valid()) {
+		return triangle_mesh;
+	}
+
+	Ref<Font> font = _get_font_or_default();
+	if (font.is_null()) {
+		return Ref<TriangleMesh>();
+	}
+
+	PoolVector<Vector3> faces;
+	faces.resize(6);
+	PoolVector<Vector3>::Write facesw = faces.write();
+
+	if (word_cache_dirty) {
+		const_cast<Label3D *>(this)->regenerate_word_cache();
+	}
+
+	int font_h = (font->get_height() + line_spacing);
+	real_t space_w = font->get_char_size(' ').width;
+	float total_h = line_count * font_h;
```

Comment:
Line spacing should not be added on the last line I think.

e.g. If there's only one line, changing the line spacing currently affects the text's rendered position.

Reply:
This makes sense. I guess `Label` have the same issue.

Reply:
Ah, I checked `Label` and this code again, it seems fine if `total_h` is used as `total_h - line_spacing` later.

So the offset when changing line spacing might be caused by other codes related to this.